---
title: "Regional association plotting of Type 2 Diabetes vs. cardiovascular diseases."
author: "[Sander W. van der Laan, PhD](https://swvanderlaan.github.io) | @swvanderlaan | s.w.vanderlaan@gmail.com"
date: "`r Sys.Date()`"
output:
  html_notebook:
    cache: yes
    code_folding: hide
    collapse: yes
    df_print: paged
    fig.align: center
    fig_caption: yes
    fig_height: 6
    fig_retina: 2
    fig_width: 7
    highlight: tango
    theme: lumen
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
mainfont: Arial
subtitle: "A 'druggable-MI-targets' project"
editor_options:
  chunk_output_type: inline
---

```{r global_options, include = FALSE}
# further define some knitr-options.
knitr::opts_chunk$set(fig.width = 12, fig.height = 8, fig.path = 'Figures/', 
                      wwarning = TRUE, # show warnings during codebook generation
  message = TRUE, # show messages during codebook generation
  error = TRUE, # do not interrupt codebook generation in case of errors,
                # usually better for debugging
  echo = TRUE,  # show R code
                      eval = TRUE)
ggplot2::theme_set(ggplot2::theme_minimal())
pander::panderOptions("table.split.table", Inf)
```

# Setup
We will clean the environment, setup the locations, define colors, and create a datestamp.

_Clean the environment._
```{r echo = FALSE}
rm(list = ls())
```

_Set locations and working directories..._
```{r LocalSystem, echo = FALSE}
### Operating System Version
### MacBook Pro
ROOT_loc = "/Users/swvanderlaan"

### MacBook Air
# ROOT_loc = "/Users/slaan3"

### Generic Locations
GENOMIC_loc = paste0(ROOT_loc, "/OneDrive - UMC Utrecht/Genomics")
STORAGE_loc = paste0(ROOT_loc, "/PLINK")

AEDB_loc = paste0(GENOMIC_loc, "/Athero-Express/AE-AAA_GS_DBs")
LAB_loc = paste0(GENOMIC_loc, "/LabBusiness")

PLINK_loc = paste0(STORAGE_loc,"")
AEGSQC_loc =  paste0(PLINK_loc, "/_AE_ORIGINALS/AEGS_COMBINED_QC2018")
MICHIMP_loc=paste0(PLINK_loc,"/_AE_ORIGINALS/AEGS_COMBINED_EAGLE2_1000Gp3v5HRCr11")

GWAS_loc = paste0(PLINK_loc,"/_GWAS_Datasets")

ASCRNA_loc = paste0(STORAGE_loc, "/_AE_ORIGINALS/AESCRNA/prepped_data")

PROJECT_loc = paste0(STORAGE_loc, "/analyses/grants/EFSD2022")
RESULTS = paste0(STORAGE_loc, "/analyses/grants/EFSD2022")

TARGET_loc = paste0(PROJECT_loc, "/targets")

### SOME VARIABLES WE NEED DOWN THE LINE
cat("\nDefining phenotypes and datasets.\n")
PROJECTNAME="RACER"

cat("\nCreate a new analysis directory, including subdirectories.\n")
# Analysis
ifelse(!dir.exists(file.path(RESULTS)), 
       dir.create(file.path(RESULTS)), 
       FALSE)
ANALYSIS_loc = paste0(RESULTS)

# Plots
ifelse(!dir.exists(file.path(ANALYSIS_loc, "/PLOTS")), 
       dir.create(file.path(ANALYSIS_loc, "/PLOTS")), 
       FALSE)
PLOT_loc = paste0(ANALYSIS_loc,"/PLOTS")

# QC plots
ifelse(!dir.exists(file.path(PLOT_loc, "/QC")), 
       dir.create(file.path(PLOT_loc, "/QC")), 
       FALSE)
QC_loc = paste0(PLOT_loc,"/QC")

# Output files
ifelse(!dir.exists(file.path(ANALYSIS_loc, "/OUTPUT")), 
       dir.create(file.path(ANALYSIS_loc, "/OUTPUT")), 
       FALSE)
OUT_loc = paste0(ANALYSIS_loc, "/OUTPUT")

cat("\nSetting working directory and listing its contents.\n")
setwd(paste0(RESULTS))
getwd()
list.files()

```

_... a package-installation function ..._
```{r}
source(paste0(PROJECT_loc, "/scripts/functions.R"))
```


_... and load those packages._
```{r loading_packages, message=FALSE, warning=FALSE}
install.packages.auto("readr")
install.packages.auto("optparse")
install.packages.auto("tools")
install.packages.auto("dplyr")
install.packages.auto("tidyr")
install.packages.auto("naniar")
install.packages.auto("pander")
install.packages.auto("R.utils")

# To get 'data.table' with 'fwrite' to be able to directly write gzipped-files
# Ref: https://stackoverflow.com/questions/42788401/is-possible-to-use-fwrite-from-data-table-with-gzfile
# install.packages("data.table", repos = "https://Rdatatable.gitlab.io/data.table")
library(data.table)

install.packages.auto("tidyverse")
install.packages.auto("knitr")
install.packages.auto("DT")
install.packages.auto("eeptools")

install.packages.auto("haven")
install.packages.auto("tableone")

install.packages.auto("BlandAltmanLeh")

# Install the devtools package from Hadley Wickham
install.packages.auto('devtools')
library(devtools) 

# for plotting
install.packages.auto("pheatmap")
install.packages.auto("forestplot")
install.packages.auto("ggplot2")
install.packages.auto("ggpubr")
install.packages.auto("ggrepel")

install.packages.auto("UpSetR")

devtools::install_github("thomasp85/patchwork")

# For regional association plots
install_github("oliviasabik/RACER") 

# Install ggrepel package if needed

library(ggrepel)

# install ggsci
install.packages.auto("ggsci")

# plotly
# install.packages.auto("plotly")

```

_We will create a datestamp and define the Utrecht Science Park Colour Scheme_.
```{r Setting: Colors}

Today = format(as.Date(as.POSIXlt(Sys.time())), "%Y%m%d")
Today.Report = format(as.Date(as.POSIXlt(Sys.time())), "%A, %B %d, %Y")

### UtrechtScienceParkColoursScheme
###
### WebsitetoconvertHEXtoRGB:http://hex.colorrrs.com.
### Forsomefunctionsyoushoulddividethesenumbersby255.
###
###	No.	Color			      HEX	(RGB)						              CHR		  MAF/INFO
###---------------------------------------------------------------------------------------
###	1	  yellow			    #FBB820 (251,184,32)				      =>	1		or 1.0>INFO
###	2	  gold			      #F59D10 (245,157,16)				      =>	2		
###	3	  salmon			    #E55738 (229,87,56)				      =>	3		or 0.05<MAF<0.2 or 0.4<INFO<0.6
###	4	  darkpink		    #DB003F ((219,0,63)				      =>	4		
###	5	  lightpink		    #E35493 (227,84,147)				      =>	5		or 0.8<INFO<1.0
###	6	  pink			      #D5267B (213,38,123)				      =>	6		
###	7	  hardpink		    #CC0071 (204,0,113)				      =>	7		
###	8	  lightpurple	    #A8448A (168,68,138)				      =>	8		
###	9	  purple			    #9A3480 (154,52,128)				      =>	9		
###	10	lavendel		    #8D5B9A (141,91,154)				      =>	10		
###	11	bluepurple		  #705296 (112,82,150)				      =>	11		
###	12	purpleblue		  #686AA9 (104,106,169)			      =>	12		
###	13	lightpurpleblue	#6173AD (97,115,173/101,120,180)	=>	13		
###	14	seablue			    #4C81BF (76,129,191)				      =>	14		
###	15	skyblue			    #2F8BC9 (47,139,201)				      =>	15		
###	16	azurblue		    #1290D9 (18,144,217)				      =>	16		or 0.01<MAF<0.05 or 0.2<INFO<0.4
###	17	lightazurblue	  #1396D8 (19,150,216)				      =>	17		
###	18	greenblue		    #15A6C1 (21,166,193)				      =>	18		
###	19	seaweedgreen	  #5EB17F (94,177,127)				      =>	19		
###	20	yellowgreen		  #86B833 (134,184,51)				      =>	20		
###	21	lightmossgreen	#C5D220 (197,210,32)				      =>	21		
###	22	mossgreen		    #9FC228 (159,194,40)				      =>	22		or MAF>0.20 or 0.6<INFO<0.8
###	23	lightgreen	  	#78B113 (120,177,19)				      =>	23/X
###	24	green			      #49A01D (73,160,29)				      =>	24/Y
###	25	grey			      #595A5C (89,90,92)				        =>	25/XY	or MAF<0.01 or 0.0<INFO<0.2
###	26	lightgrey		    #A2A3A4	(162,163,164)			      =>	26/MT
### 
###	ADDITIONAL COLORS
###	27	midgrey			#D7D8D7
###	28	verylightgrey	#ECECEC"
###	29	white			#FFFFFF
###	30	black			#000000
###----------------------------------------------------------------------------------------------

uithof_color = c("#FBB820","#F59D10","#E55738","#DB003F","#E35493","#D5267B",
                 "#CC0071","#A8448A","#9A3480","#8D5B9A","#705296","#686AA9",
                 "#6173AD","#4C81BF","#2F8BC9","#1290D9","#1396D8","#15A6C1",
                 "#5EB17F","#86B833","#C5D220","#9FC228","#78B113","#49A01D",
                 "#595A5C","#A2A3A4", "#D7D8D7", "#ECECEC", "#FFFFFF", "#000000")

uithof_color_legend = c("#FBB820", "#F59D10", "#E55738", "#DB003F", "#E35493",
                        "#D5267B", "#CC0071", "#A8448A", "#9A3480", "#8D5B9A",
                        "#705296", "#686AA9", "#6173AD", "#4C81BF", "#2F8BC9",
                        "#1290D9", "#1396D8", "#15A6C1", "#5EB17F", "#86B833",
                        "#C5D220", "#9FC228", "#78B113", "#49A01D", "#595A5C",
                        "#A2A3A4", "#D7D8D7", "#ECECEC", "#FFFFFF", "#000000")
### ----------------------------------------------------------------------------
```

# Introduction

We will parse the data to create regional association plots for each of the 11 loci. 

# Setting the NPG colors

```{r}
library("scales")
pal_npg("nrc")(10)
show_col(pal_npg("nrc")(10))

# show_col(pal_npg("nrc", alpha = 0.6)(10))

```

# show_col(pal_npg("nrc", alpha = 0.6)(10))

# Load data

We need to load the data first.

```{r}
# when we need to load the data

# gwas_sumstats <- readRDS(file = paste0(OUT_loc, "/gwas_sumstats_complete.rds"))

gwas_t2d <- fread(paste0(GWAS_loc,"/_cojo/rsid/T2D.cojo.gz"),
                         showProgress = TRUE)

gwas_cad <- fread(paste0(GWAS_loc,"/_cojo/rsid/CAD.cojo.gz"),
                         showProgress = TRUE)

gwas_as <- fread(paste0(GWAS_loc,"/_cojo/rsid/AS.cojo.gz"),
                         showProgress = TRUE)

gwas_cimt <- fread(paste0(GWAS_loc,"/_cojo/rsid/CIMT.cojo.gz"),
                         showProgress = TRUE)

```

```{r}

saveRDS(gwas_t2d, file = paste0(OUT_loc, "/gwas_sumstats_t2d.rds"))
saveRDS(gwas_cad, file = paste0(OUT_loc, "/gwas_sumstats_cad.rds"))
saveRDS(gwas_as, file = paste0(OUT_loc, "/gwas_sumstats_as.rds"))
saveRDS(gwas_cimt, file = paste0(OUT_loc, "/gwas_sumstats_cimt.rds"))
rm(gwas_t2d, gwas_cad, gwas_as, gwas_cimt)

```

```{r}


gwas_cad_racert <- subset(gwas_cad,
                              select = c("SNP", "p"))

gwas_as_racert <- subset(gwas_as,
                              select = c("SNP", "p"))

gwas_cimt_racert <- subset(gwas_cimt,
                              select = c("SNP", "p"))

rm(gwas_t2d, gwas_cad, gwas_as, gwas_cimt)

```

```{r}
# ref_1kGp3 <- fread(paste0(PLINK_loc, "/references/1kGp3.nonbia.sumstats.newids.chrbp.txt.gz"),
#                    showProgress = TRUE)
# saveRDS(ref_1kGp3, file = paste0(OUT_loc, "/1kGp3.nonbia.sumstats.newids.chrbp.rds"))
ref_1kGp3 <- readRDS(file = paste0(OUT_loc, "/1kGp3.nonbia.sumstats.newids.chrbp.rds"))

```


```{r}
gwas_t2d <- readRDS(paste0(OUT_loc, "/gwas_sumstats_t2d.rds"))
gwas_t2d_racert <- subset(gwas_t2d,
                              select = c("SNP", "p"))
rm(gwas_t2d)
gwas_t2d_racer <- merge(gwas_t2d_racert, ref_1kGp3, by.x = "SNP", by.y = "ID",
                        all.x = TRUE, sort = FALSE)
rm(gwas_t2d_racert)

saveRDS(gwas_t2d_racer, file = paste0(OUT_loc, "/gwas_t2d_racer.rds"))

# gwas_t2d_racer <- readRDS(paste0(OUT_loc, "/gwas_t2d_racer.rds"))

```

```{r}
gwas_cad <- readRDS(paste0(OUT_loc, "/gwas_sumstats_cad.rds"))
gwas_cad_racert <- subset(gwas_cad,
                              select = c("SNP", "p"))
rm(gwas_cad)
gwas_cad_racer <- merge(gwas_cad_racert, ref_1kGp3, by.x = "SNP", by.y = "ID",
                        all.x = TRUE, sort = FALSE)
rm(gwas_cad_racert)

saveRDS(gwas_cad_racer, file = paste0(OUT_loc, "/gwas_cad_racer.rds"))

# gwas_cad_racer <- readRDS(paste0(OUT_loc, "/gwas_cad_racer.rds"))

```

```{r}
gwas_as <- readRDS(paste0(OUT_loc, "/gwas_sumstats_as.rds"))
gwas_as_racert <- subset(gwas_as,
                              select = c("SNP", "p"))
rm(gwas_as)
gwas_as_racer <- merge(gwas_as_racert, ref_1kGp3, by.x = "SNP", by.y = "ID",
                        all.x = TRUE, sort = FALSE)
rm(gwas_as_racert)

saveRDS(gwas_as_racer, file = paste0(OUT_loc, "/gwas_as_racer.rds"))

# gwas_as_racer <- readRDS(paste0(OUT_loc, "/gwas_as_racer.rds"))

```

```{r}
gwas_cimt <- readRDS(paste0(OUT_loc, "/gwas_sumstats_cimt.rds"))
gwas_cimt_racert <- subset(gwas_cimt,
                              select = c("SNP", "p"))
rm(gwas_cimt)
gwas_cimt_racer <- merge(gwas_cimt_racert, ref_1kGp3, by.x = "SNP", by.y = "ID",
                        all.x = TRUE, sort = FALSE)
rm(gwas_cimt_racert)

saveRDS(gwas_cimt_racer, file = paste0(OUT_loc, "/gwas_cimt_racer.rds"))

# gwas_cimt_racer <- readRDS(paste0(OUT_loc, "/gwas_cimt_racer.rds"))

```

```{r}
rm(ref_1kGp3)
```

# Regional association plotting

## Intersected genes

We are interested in the genes identified in [Slenders et al.](https://academic.oup.com/ehjopen/article/doi/10.1093/ehjopen/oeab043/6472573). 

```{r}
library(openxlsx)
target_genes <- read.xlsx(paste0(TARGET_loc, "/20220223_targets.xlsx"))

DT::datatable(target_genes)

```

## Gene-based associated genes

```{r}
T2Dmagma <- as_tibble(fread(paste0(STORAGE_loc, "/_GWAS_Datasets/_magma/T2D/T2D.magma.genes.out")))

BMImagma <- fread(paste0(STORAGE_loc, "/_GWAS_Datasets/_magma/BMI/BMI.magma.genes.out"))

ASmagma <- fread(paste0(STORAGE_loc, "/_GWAS_Datasets/_magma/AS/AS.magma.genes.out"))
ISmagma <- fread(paste0(STORAGE_loc, "/_GWAS_Datasets/_magma/IS/IS.magma.genes.out"))
SVDmagma <- fread(paste0(STORAGE_loc, "/_GWAS_Datasets/_magma/SVD/SVD.magma.genes.out"))
CESmagma <- fread(paste0(STORAGE_loc, "/_GWAS_Datasets/_magma/CES/CES.magma.genes.out"))
LASmagma <- fread(paste0(STORAGE_loc, "/_GWAS_Datasets/_magma/LAS/LAS.magma.genes.out"))

CADmagma <- fread(paste0(STORAGE_loc, "/_GWAS_Datasets/_magma/CAD/CAD.magma.genes.out"))

cIMTmagma <- fread(paste0(STORAGE_loc, "/_GWAS_Datasets/_magma/cIMT/cIMT.magma.genes.out"))

```

### Extract significant genes

```{r}

p_adj_threshold_t2d = 0.05/nrow(T2Dmagma)
target_genes_t2d <- unique(subset(T2Dmagma, T2Dmagma$P < p_adj_threshold_t2d, select = c("SYMBOL"))$SYMBOL)
length(target_genes_t2d)


p_adj_threshold_bmi = 0.05/nrow(BMImagma)
target_genes_bmi <- unique(subset(BMImagma, BMImagma$P < p_adj_threshold_bmi, select = c("SYMBOL"))$SYMBOL)
length(target_genes_bmi)


p_adj_threshold_as = 0.05/nrow(ASmagma)
target_genes_as <- unique(subset(ASmagma, ASmagma$P < p_adj_threshold_as, select = c("SYMBOL"))$SYMBOL)
length(target_genes_as)

p_adj_threshold_is = 0.05/nrow(ISmagma)
target_genes_is <- unique(subset(ISmagma, ISmagma$P < p_adj_threshold_is, select = c("SYMBOL"))$SYMBOL)
length(target_genes_is)

p_adj_threshold_svd = 0.05/nrow(SVDmagma)
target_genes_svd <- unique(subset(SVDmagma, SVDmagma$P < p_adj_threshold_svd, select = c("SYMBOL"))$SYMBOL)
length(target_genes_svd)

p_adj_threshold_ces = 0.05/nrow(CESmagma)
target_genes_ces <- unique(subset(CESmagma, CESmagma$P < p_adj_threshold_ces, select = c("SYMBOL"))$SYMBOL)
length(target_genes_ces)

p_adj_threshold_las = 0.05/nrow(LASmagma)
target_genes_las <- unique(subset(LASmagma, LASmagma$P < p_adj_threshold_las, select = c("SYMBOL"))$SYMBOL)
length(target_genes_las)


p_adj_threshold_cad = 0.05/nrow(CADmagma)
target_genes_cad <- unique(subset(CADmagma, CADmagma$P < p_adj_threshold_cad, select = c("SYMBOL"))$SYMBOL)
length(target_genes_cad)


p_adj_threshold_cimt = 0.05/nrow(cIMTmagma)
target_genes_cimt <- unique(subset(cIMTmagma, cIMTmagma$P < p_adj_threshold_cimt, select = c("SYMBOL"))$SYMBOL)
length(target_genes_cimt)

```

### Overlap between diseases

What is the 1-to-1 overlap between diseases?

```{r}
t2d_vs_bmi <- match(target_genes_t2d, target_genes_bmi)

t2d_vs_as <- match(target_genes_t2d, target_genes_as)
t2d_vs_is <- match(target_genes_t2d, target_genes_is)
t2d_vs_svd <- match(target_genes_t2d, target_genes_svd)
t2d_vs_las <- match(target_genes_t2d, target_genes_las)
t2d_vs_ces <- match(target_genes_t2d, target_genes_ces)

t2d_vs_cad <- match(target_genes_t2d, target_genes_cad)
t2d_vs_cimt <- match(target_genes_t2d, target_genes_cimt)


cat("*T2D vs BMI (positive control)\n")
target_genes_t2d[!is.na(t2d_vs_bmi)]
targets_t2d_vs_bmi <- target_genes_t2d[!is.na(t2d_vs_bmi)]

cat("*T2D vs all stroke\n")
target_genes_t2d[!is.na(t2d_vs_as)]
targets_t2d_vs_as <- target_genes_t2d[!is.na(t2d_vs_as)]

cat("*T2D vs ischemic stroke\n")
target_genes_t2d[!is.na(t2d_vs_is)]
targets_t2d_vs_is <- target_genes_t2d[!is.na(t2d_vs_is)]

cat("*T2D vs small vessel disease\n")
target_genes_t2d[!is.na(t2d_vs_svd)]
targets_t2d_vs_svd <- target_genes_t2d[!is.na(t2d_vs_svd)]

cat("*T2D vs large artery stroke\n")
target_genes_t2d[!is.na(t2d_vs_las)]
targets_t2d_vs_las <- target_genes_t2d[!is.na(t2d_vs_las)]

cat("*T2D vs cardioembolic stroke\n")
target_genes_t2d[!is.na(t2d_vs_ces)]
targets_t2d_vs_ces <- target_genes_t2d[!is.na(t2d_vs_ces)]

cat("*T2D vs CAD\n")
target_genes_t2d[!is.na(t2d_vs_cad)]
targets_t2d_vs_cad <- target_genes_t2d[!is.na(t2d_vs_cad)]

cat("*T2D vs carotid IMT\n")
target_genes_t2d[!is.na(t2d_vs_cimt)]
targets_t2d_vs_cimt <- target_genes_t2d[!is.na(t2d_vs_cimt)]

rm(t2d_vs_bmi, t2d_vs_as, t2d_vs_is, t2d_vs_svd,  t2d_vs_las, t2d_vs_ces, t2d_vs_cad, t2d_vs_cimt)

```



### T2D vs AS
```{r}
library(RACER)

# Make directory for plots
ifelse(!dir.exists(file.path(PROJECT_loc, "/RACER")), 
       dir.create(file.path(PROJECT_loc, "/RACER")), 
       FALSE)
RACER_loc = paste0(PROJECT_loc,"/RACER")

ifelse(!dir.exists(file.path(RACER_loc, "/AS")), 
       dir.create(file.path(RACER_loc, "/AS")), 
       FALSE)
AS_RACER_loc = paste0(RACER_loc,"/AS")

for(GENE in targets_t2d_vs_as){
  cat(paste0("Getting data for ", GENE,".\n"))
  target=GENE
  tempCHR <- subset(T2Dmagma, SYMBOL == target)$CHR
  tempSTART <- subset(T2Dmagma, SYMBOL == target)$START
  tempEND <- subset(T2Dmagma, SYMBOL == target)$STOP
  tempENSEMBLnr <- subset(T2Dmagma, SYMBOL == target)$GENE

  cat("\nSubset required data.\n")
  temp <- subset(gwas_t2d_racer, CHROM == tempCHR & (POS >= tempSTART & POS <= tempEND))
  temp_as <- subset(gwas_as_racer, CHROM == tempCHR & (POS >= tempSTART & POS <= tempEND))
  
  temp$CHROM <- as.numeric(temp$CHROM)
  temp$POS <- as.numeric(temp$POS)
  temp$p <- as.numeric(temp$p)
  
  cat("\nFormatting association data.\n")
  temp_f = RACER::formatRACER(assoc_data = temp, chr_col = 3, pos_col = 4, p_col = 2)
  temp_as_f = RACER::formatRACER(assoc_data = temp_as, chr_col = 3, pos_col = 4, p_col = 2)

  cat("\nGetting LD data.\n")
  temp_f_ld = RACER::ldRACER(assoc_data = temp_f, rs_col = 1, pops = "EUR", auto_snp = TRUE)
  lead_variant <- subset(temp_f_ld, LABEL == "LEAD")$RS_ID
  
  temp_as_f_ld = RACER::ldRACER(assoc_data = temp_as_f, rs_col = 1, pops = "EUR", lead_snp = lead_variant)

  cat(paste0("\nPlotting region surrounding ", GENE," on ",tempCHR,":",tempSTART,"-",tempEND,".\n"))
  
  p1 <- mirrorPlotRACER(assoc_data1 = temp_f_ld, 
                        assoc_data2 = temp_as_f_ld, chr = tempCHR, 
                        plotby = "coord", start_plot = tempSTART, end_plot = tempEND,
                        # name1 = paste0("Type 2 Diabetes (", lead_variant,")"),
                        name1 = "Type 2 Diabetes",
                        name2 = "All stroke", 
                        label_lead = TRUE) 

  print(p1 )
  cat(paste0("Saving image for ", GENE,".\n"))
  ggsave(filename = paste0(AS_RACER_loc, "/", tempENSEMBLnr, ".", Today, ".",GENE,".regional_assoc.t2d_vs_as.png"), plot = last_plot())
  ggsave(filename = paste0(AS_RACER_loc, "/", tempENSEMBLnr, ".", Today, ".",GENE,".regional_assoc.t2d_vs_as.pdf"), plot = last_plot())
  ggsave(filename = paste0(AS_RACER_loc, "/", tempENSEMBLnr, ".", Today, ".",GENE,".regional_assoc.t2d_vs_as.eps"), plot = last_plot())

  # rm(temp, p1, 
  #    temp_f, temp_f_ld,
  #    tempCHR, tempSTART, tempEND,
  #    tempENSEMBLnr,
  #    temp_as, temp_as_f, temp_as_f_ld,
  #    GENE)
  
}
```


### T2D vs cIMT

We are interested in those genes expressed in monocytes or macrophages, endothelial (ECs) and smooth muscle cells (SMCs). After inspection of the single-cell expression profiles, we identified the following genes for cIMT:

- _SOX7_ --> ECs - we use `ENSG00000171056` 
- _BLK_ --> monocytes - `ENSG00000136573`
- _PPP1R3B_ --> SMCs - `ENSG00000173281`


```{r}
library(RACER)

# Make directory for plots
ifelse(!dir.exists(file.path(PROJECT_loc, "/RACER")), 
       dir.create(file.path(PROJECT_loc, "/RACER")), 
       FALSE)
RACER_loc = paste0(PROJECT_loc,"/RACER")

ifelse(!dir.exists(file.path(RACER_loc, "/CIMT")), 
       dir.create(file.path(RACER_loc, "/CIMT")), 
       FALSE)
CIMT_RACER_loc = paste0(RACER_loc,"/CIMT")

# targets_t2d_vs_cimt_short <- c("SOX7", "BLK", "PPP1R3B") 
targets_t2d_vs_cimt_short <- c("ENSG00000171056", "ENSG00000136573", "ENSG00000173281") 

# source(paste0(PROJECT_loc, "/scripts/functions.R"))
for(GENE in targets_t2d_vs_cimt_short){
  cat(paste0("Getting data for ", GENE,".\n"))
  target=GENE
  tempCHR <- subset(T2Dmagma, GENE == target)$CHR
  tempSTART <- subset(T2Dmagma, GENE == target)$START
  tempEND <- subset(T2Dmagma, GENE == target)$STOP
  tempENSEMBLnr <- subset(T2Dmagma, GENE == target)$SYMBOL

  cat("\nSubset required data.\n")
  temp <- subset(gwas_t2d_racer, CHROM == tempCHR & (POS >= tempSTART & POS <= tempEND))
  temp_cimt <- subset(gwas_cimt_racer, CHROM == tempCHR & (POS >= tempSTART & POS <= tempEND))
  
  temp$CHROM <- as.numeric(temp$CHROM)
  temp$POS <- as.numeric(temp$POS)
  temp$p <- as.numeric(temp$p)
  
  cat("\nFormatting association data.\n")
  temp_f = RACER::formatRACER(assoc_data = temp, chr_col = 3, pos_col = 4, p_col = 2)
  temp_cimt_f = RACER::formatRACER(assoc_data = temp_cimt, chr_col = 3, pos_col = 4, p_col = 2)

  cat("\nGetting LD data.\n")
  temp_f_ld = RACER::ldRACER(assoc_data = temp_f, rs_col = 1, pops = "EUR", auto_snp = TRUE)
  temp_cimt_f_ld = RACER::ldRACER(assoc_data = temp_cimt_f, rs_col = 1, pops = "EUR", auto_snp = TRUE)
  # print(head(temp_cimt_f_ld))
  cat(paste0("\nPlotting region surrounding ", GENE," on ",tempCHR,":",tempSTART,"-",tempEND,".\n"))
  p1 <- mirrorPlotRACER(assoc_data1 = temp_f_ld, 
                        assoc_data2 = temp_cimt_f_ld, chr = tempCHR, 
                        plotby = "coord", start_plot = tempSTART, end_plot = tempEND,
                        name1 = "Type 2 Diabetes",
                        name2 = "Carotid intima-media thickness", 
                        label_lead = TRUE) 

  print(p1 )
  cat(paste0("Saving image for ", GENE,".\n"))
  ggsave(filename = paste0(CIMT_RACER_loc, "/", tempENSEMBLnr, ".", Today, ".",GENE,".regional_assoc.t2d_vs_cimt.png"), plot = last_plot())
  # ggsave(filename = paste0(CIMT_RACER_loc, "/", tempENSEMBLnr, ".", Today, ".",GENE,".regional_assoc.t2d_vs_cimt.pdf"), plot = last_plot())
  # ggsave(filename = paste0(CIMT_RACER_loc, "/", tempENSEMBLnr, ".", Today, ".",GENE,".regional_assoc.t2d_vs_cimt.eps"), plot = last_plot())

  rm(temp, p1,
     temp_f, temp_f_ld,
     tempCHR, tempSTART, tempEND,
     tempENSEMBLnr,
     temp_cimt, temp_cimt_f, temp_cimt_f_ld,
     GENE)
  
}
```


### T2D vs CAD

```{r}
library(RACER)

# Make directory for plots
ifelse(!dir.exists(file.path(PROJECT_loc, "/RACER")), 
       dir.create(file.path(PROJECT_loc, "/RACER")), 
       FALSE)
RACER_loc = paste0(PROJECT_loc,"/RACER")

ifelse(!dir.exists(file.path(RACER_loc, "/CAD")), 
       dir.create(file.path(RACER_loc, "/CAD")), 
       FALSE)
CAD_RACER_loc = paste0(RACER_loc,"/CAD")

for(GENE in targets_t2d_vs_cad){
  cat(paste0("Getting data for ", GENE,".\n"))
  target=GENE
  tempCHR <- subset(T2Dmagma, SYMBOL == target)$CHR
  tempSTART <- subset(T2Dmagma, SYMBOL == target)$START
  tempEND <- subset(T2Dmagma, SYMBOL == target)$STOP
  tempENSEMBLnr <- subset(T2Dmagma, SYMBOL == target)$GENE

  cat("\nSubset required data.\n")
  temp <- subset(gwas_t2d_racer, CHROM == tempCHR & (POS >= tempSTART & POS <= tempEND))
  temp_cad <- subset(gwas_cad_racer, CHROM == tempCHR & (POS >= tempSTART & POS <= tempEND))

  temp$CHROM <- as.numeric(temp$CHROM)
  temp$POS <- as.numeric(temp$POS)
  temp$p <- as.numeric(temp$p)
  
  cat("\nFormatting association data.\n")
  temp_f = RACER::formatRACER(assoc_data = temp, chr_col = 3, pos_col = 4, p_col = 2)
  temp_cad_f = RACER::formatRACER(assoc_data = temp_cad, chr_col = 3, pos_col = 4, p_col = 2)

  cat("\nGetting LD data.\n")
  temp_f_ld = RACER::ldRACER(assoc_data = temp_f, rs_col = 1, pops = "EUR", auto_snp = TRUE)
  temp_cad_f_ld = RACER::ldRACER(assoc_data = temp_cad_f, rs_col = 1, pops = "EUR", auto_snp = TRUE)

  cat(paste0("\nPlotting region surrounding ", GENE," on ",tempCHR,":",tempSTART,"-",tempEND,".\n"))
  source(paste0(PROJECT_loc, "/scripts/functions.R"))
  p1 <- mirrorPlotRACER(assoc_data1 = temp_f_ld, 
                        assoc_data2 = temp_cad_f_ld, chr = tempCHR, 
                        plotby = "coord", start_plot = tempSTART, end_plot = tempEND,
                        name1 = "Type 2 Diabetes",
                        name2 = "Coronary artery disease", 
                        label_lead = TRUE) 

  print(p1 )
  cat(paste0("Saving image for ", GENE,".\n"))
  ggsave(filename = paste0(CAD_RACER_loc, "/", tempENSEMBLnr, ".", Today, ".",GENE,".regional_assoc.t2d_vs_cad.png"), plot = last_plot())
  # ggsave(filename = paste0(CAD_RACER_loc, "/", tempENSEMBLnr, ".", Today, ".",GENE,".regional_assoc.t2d_vs_cad.pdf"), plot = last_plot())
  # ggsave(filename = paste0(CAD_RACER_loc, "/", tempENSEMBLnr, ".", Today, ".",GENE,".regional_assoc.t2d_vs_cad.eps"), plot = last_plot())

  rm(temp, p1, 
     temp_f, temp_f_ld,
     tempCHR, tempSTART, tempEND,
     tempENSEMBLnr,
     temp_cad, temp_cad_f, temp_cad_f_ld,
     GENE)
  
}


```



# Session information

------

    Version:      v1.0.0
    Last update:  2022-02-28
    Written by:   Sander W. van der Laan (s.w.vanderlaan-2[at]umcutrecht.nl).
    Description:  Script to create plot regional association plots.
    Minimum requirements: R version 3.4.3 (2017-06-30) -- 'Single Candle', Mac OS X El Capitan
    
    Changes log
    * v1.0.0 Initial version. 

------

```{r eval = TRUE}
sessionInfo()
```


# Saving environment
```{r Saving}

save.image(paste0(PROJECT_loc, "/",Today,".",PROJECTNAME,".RegionalAssociationPlots.RData"))
```


------
<sup>&copy; 1979-2022 Sander W. van der Laan | s.w.vanderlaan[at]gmail.com | [swvanderlaan.github.io](https://swvanderlaan.github.io).</sup>
------

  
